File: solutions/clojure/allergies/1/src/allergies.clj
(ns allergies)

(def allergies-map
  {
     :eggs 0
     :peanuts 1
     :shellfish 2
     :strawberries 3
     :tomatoes 4
     :chocolate 5
     :pollen 6
     :cats 7
  })

(defn allergic-to? [code key]
  (->> (key allergies-map)
    (bit-shift-right code)
    (bit-and 1)
    (= 1)))

(defn allergies [code]
  (filter #(allergic-to? code %) (keys allergies-map)))

File: solutions/nim/allergies/1/allergies.nim
import std/tables

type
  Allergen* = enum
    Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats

let
  allergyScoreMap = [
       (Eggs, 1),
       (Peanuts, 2),
       (Shellfish, 4),
       (Strawberries, 8),
       (Tomatoes, 16),
       (Chocolate, 32),
       (Pollen, 64),
       (Cats, 128) 
  ].toOrderedTable

proc isAllergicTo*(score: int, allergen: Allergen): bool =
  let code = allergyScoreMap[allergen]
  (code and score) == code

proc allergies*(score: int): set[Allergen] =
  var
    result: set[Allergen] = {}
  for k in allergyScoreMap.keys:
    if isAllergicTo(score, k):
        result = result + {k}
  result
